Fix node e2e issues on selinux enabled systems

aveshagarwal opened the pull request:
It fixes following 3 node e2es:

```
[Fail] [k8s.io] Container Runtime Conformance Test container runtime conformance blackbox test when starting a container that exits [It] it should run with the expected status [Conformance] 
/root/upstream-code/gocode/src/k8s.io/kubernetes/test/e2e_node/runtime_conformance_test.go:114

[Fail] [k8s.io] Kubelet metrics api when querying /stats/summary [It] it should report resource usage through the stats api 
/root/upstream-code/gocode/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:158
```

```
[Fail] [k8s.io] Container Runtime Conformance Test container runtime conformance blackbox test when starting a container that exits [It] should report termination message if TerminationMessagePath is set [Conformance]
/root/upstream-code/gocode/src/k8s.io/kubernetes/test/e2e_node/runtime_conformance_test.go:150
```

@kubernetes/rh-cluster-infra

aveshagarwal commented:
@eparis @pmorie

eparis commented:
This does the right thing on non-selinux systems?

eparis commented on test/e2e_node/runtime_conformance_test.go:
```diff
@@ -45,8 +45,12 @@ var _ = framework.KubeDescribe("Container Runtime Conformance Test", func() {
 					Image: ImageRegistry[busyBoxImage],
+					SecurityContext: &api.SecurityContext{
+						Privileged: &priv,
+					},
```
Why do "we" need 'priv' and others don't?

aveshagarwal replied:
In this test, data is being written to a file in /tmp, so privileged.

However, i think that another way would be to use emptydir with selinux labels, and it should work across container restart (as this test checks for restarts) I think.

aveshagarwal commented:
@eparis I haven't checked non selinux system.

timstclair commented on test/e2e_node/kubelet_test.go:
```diff
@@ -204,7 +204,12 @@ func createPod(f *framework.Framework, podName string, containers []api.Containe
-			Volumes:       volumes,
+			SecurityContext: &api.PodSecurityContext{
+				SELinuxOptions: &api.SELinuxOptions{
+					Level: "s0:c123,c456",
```
Are these references specific to a particular SELinux setup, or valid for all SELinux setups? Are they documented anywhere?

Commit:
Fix selinux issue with the following node e2e test:
[Fail] [k8s.io] Container Runtime Conformance Test container runtime conformance blackbox test when starting a container that exits [It] it should run with the expected status [Conformance]
/root/upstream-code/gocode/src/k8s.io/kubernetes/test/e2e_node/runtime_conformance_test.go:114

aveshagarwal commented:
@timstclair it is general and should work with any selinux setup.

aveshagarwal commented:
@eparis I have reworked the implementation of the test case to use emptydir and not /tmp with hostPath. So now the container is not created with privileged true.

eparis commented:
There are selinux setups with NO level component. Also the number of categories are also configurable. Can we just use c0:c1.  We'll assume they have at least 2 cats?

aveshagarwal commented:
@eparis, I think for these tests, s0 should be enough, and not worry about mcs labels?

Commit:
Fix following node e2e issue on selinux enabled systems:

[Fail] [k8s.io] Container Runtime Conformance Test container runtime conformance blackbox test when starting a container that exits [It] should report termination message if TerminationMessagePath is set [Conformance]
/root/upstream-code/gocode/src/k8s.io/kubernetes/test/e2e_node/runtime_conformance_test.go:150

aveshagarwal commented:
@pmorie  @eparis @timstclair  feedback?

timstclair commented on test/e2e_node/container.go:
```diff
@@ -64,6 +64,37 @@ func (cc *ConformanceContainer) Create() error {
+			Volumes:          cc.Volumes,
+			ImagePullSecrets: imagePullSecrets,
+		},
+	}
```
Lets move from here up to a helper method so it's not duplicated.

timstclair replied:
I mean except the SecurityContext, which you can add after.

aveshagarwal commented:
@timstclair PTAL

timstclair commented:
LGTM

aveshagarwal committed:
Fix selinux issue with following node e2e test:
[Fail] [k8s.io] Kubelet metrics api when querying /stats/summary [It] it should report resource usage through the stats api
/root/upstream-code/gocode/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:158

timstclair commented:
Still LGTM, but please squash your commits.

aveshagarwal commented:
@timstclair  I did it intentionally to have each commit for each issue. So that it is easier to see what changes were done for each of 3 issues separately. Is there any reason to not have 3 commits?

timstclair commented:
Oh yeah, that's fine. I thought they were for revisions to the PR.

aveshagarwal commented:
@k8s-bot node e2e test this issue: #IGNORE

aveshagarwal commented:
@k8s-bot node e2e test this issue: #IGNORE

aveshagarwal commented:
@k8s-bot node e2e test this issue: #IGNORE

Random-Liu commented:
@aveshagarwal Is the the PodSecurityContext configuration always needed on SELinux? Or only when we use some features, we need that?

Currently, we have a unique function to create pod in test, and we already munge pod spec there, https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pods.go#L83.

Is it possible to apply the selinux option there, when the operating system is selinux. Or else spread the logic in the test is really confusing.

smarterclayton commented:
The termination message path test had privileged set, but that masked a real problem - termination message paths didn't work on SELinux systems, or when non-root users tried to edit it.  Will be fixing as part of https://github.com/kubernetes/community/issues/154

smarterclayton commented:
Derek actually fixed that in 42ff2b3890f9fa3e8e5898b9bf81699db4744ad4
